### PR TITLE
add RENAME DATABASE

### PR DESCRIPTION
**Pull request by linearb**
Support for RENAME DATABASE as in #4154 

Comments welcome.

CQs not dealt with yet-- advice appreciated,
and not sure if correct proto version generated from meta.proto.

### CODE REVIEW

**Comment by otoolep** [comment]
Thanks @linearb -- I think there are more changes required to the TSDB layer. To see this you should add tests that query for data in a database, get results back, rename the database, and run the exact same query against the renamed database. You should get the same data back.

Make sense?

**Comment by linearb** [comment]
Thanks @otoolep . Adding that test is a good suggestion.

Are you sure TSDB changes are needed?--- it works when I do such a test by hand:

curl -G 'http://localhost:8086/query' --data-urlencode "q=CREATE DATABASE db_old"
{"results":[{}]}

curl -XPOST 'http://localhost:8086/write?db=db_old' -d 'cpu,host=server01,region=useast load=4 1434055565000000000'

curl -XPOST 'http://localhost:8086/write?db=db_old' -d 'cpu,host=server01,region=useast load=23 1434055569000000000'

curl -G http://localhost:8086/query?pretty=true --data-urlencode "db=db_old" --data-urlencode "q=SELECT * FROM cpu WHERE region='useast'"
{ "results": [ { "series": [ { "name": "cpu", "columns": [ "time", "host", "load", "region" ], "values": [ [ "2015-06-11T20:46:05Z", "server01", 4, "useast" ], [ "2015-06-11T20:46:09Z", "server01", 23, "useast" ] ] } ] } ] }

curl -G http://localhost:8086/query?pretty=true --data-urlencode "db=db_new" --data-urlencode "q=SELECT * FROM cpu WHERE region='us east'"
{"results":[{"error": "database not found: db_new"}]}

curl -G http://localhost:8086/query?pretty=true --data-urlencode "q=RENAME DATABASE db_old TO db_new"
{"results": [{}]}

curl -G http://localhost:8086/query?pretty=true --data-urlencode "db=db_new" --data-urlencode "q=SELECT * FROM cpu WHERE region='useast'"
{ "results": [ { "series": [ { "name": "cpu", "columns": [ "time", "host", "load", "region" ], "values": [ [ "2015-06-11T20:46:05Z", "server01", 4, "useast" ], [ "2015-06-11T20:46:09Z", "server01", 23, "useast" ] ] } ] } ] }

curl -G http://localhost:8086/query?pretty=true --data-urlencode "db=db_old" --data-urlencode "q=SELECT * FROM cpu WHERE region='useast'"
{ "results": [ { "error": "database not found: db_old" } ] }

**Comment by otoolep** [comment]
OK, I guess TSDB layer changes may not be required.

We still need that test however. Can you add it? We'll also need to run this command by the team.

**Comment by linearb** [comment]
Sure, I can add the test.

**Comment by gunnaraasen** [comment]
I'm not a fan of the `RENAME DATABASE` syntax. `ALTER DATABASE` feels more SQL-like. The full syntax would be something like:

```
ALTER DATABASE <old_db_name> RENAME TO <new_db_name>
```

Which roughly follows other SQL database's renaming syntax and complements the existing `ALTER RETENTION POLICY` statement. Just my two cents.

**Comment by beckettsean** [comment]
+1 for `ALTER DATABASE`. That fits in with our other `ALTER USER` and `ALTER RETENTION POLICY` queries.

**Comment by otoolep** [comment]
Yeah, good call, it should be `ALTER`.

**Comment by linearb** [comment]
OK - ALTER it is.
Interestingly, mysql once had a RENAME DATABASE that was considered dangerous:
https://dev.mysql.com/doc/refman/5.1/en/rename-database.html

**Comment by linearb** [comment]
Realized that user privileges also needed updating on db rename. 
So added something for that also.

Still unsure how to update CQs correctly, but if that's going to be
complicated perhaps better to split that off as a separate issue.

CHANGELOG.md updated
mergeable
Tests pass
Signed CLA

File: cmd/influxd/run/server_test.go [hunk 1]
```diff
@@ -67,17 +67,42 @@ func TestServer_DatabaseCommands(t *testing.T) {
 				exp:     `{"results":[{"series":[{"name":"databases","columns":["name"],"values":[["db0"],["db1"]]}]}]}`,
 			},
 			&Query{
+				name:    "rename database should succeed",
```
Comment (otoolep) [inline cmd/influxd/run/server_test.go]: Should read `alter database`.

Comment (otoolep) [inline cmd/influxd/run/server_test.go]: Actuall, I don't think it matters, now that I look at the tests.

File: influxql/parser.go [hunk 2]
```diff
@@ -219,16 +219,22 @@ func (p *Parser) parseDropStatement() (Statement, error) {
...
 		return p.parseAlterRetentionPolicyStatement()
+	case DATABASE:
+		return p.parseAlterDatabaseRenameStatement()
 	}
 
-	return nil, newParseError(tokstr(tok, lit), []string{"RETENTION"}, pos)
+	return nil, newParseError(tokstr(tok, lit), []string{"RETENTION", "DATABASE"}, pos)
 }
 
+// parseDropStatement parses a string and returns a drop statement.
```
Comment (otoolep) [inline influxql/parser.go]: Why is this here?

File: influxql/parser_test.go [hunk 3]
```diff
@@ -1412,12 +1412,19 @@ func TestParser_ParseStatement(t *testing.T) {
 			s:    `ALTER RETENTION POLICY policy1 ON testdb REPLICATION 4`,
 			stmt: newAlterRetentionPolicyStatement("policy1", "testdb", -1, 4, false),
 		},
+
```
Comment (otoolep) [inline influxql/parser_test.go]: Can you remove this blank line?

**Comment by otoolep** [comment]
@linearb -- this looks really great, just some minor feedback. All community-contributions require 2 reviews by core developers, so we just need to wait for that too.

@dgnorton -- can you also review?

File: meta/internal/meta.proto [hunk 4]
```diff
@@ -105,6 +105,7 @@ message Command {
 		SetDataCommand                   = 17;
 		SetAdminPrivilegeCommand         = 18;
 		UpdateNodeCommand                = 19;
+                RenameDatabaseCommand            = 20;
```
Comment (otoolep) [inline meta/internal/meta.proto]: Indentation looks slightly off here.

**Comment by otoolep** [comment]
Great test coverage, +1.

File: influxql/parser.go [hunk 5]
```diff
@@ -1352,6 +1358,44 @@ func (p *Parser) parseDropDatabaseStatement() (*DropDatabaseStatement, error) {
...
+		return nil, newParseError(tokstr(tok, lit), []string{"RENAME"}, pos)
+	}
+
+	// Parse TO clause.
+	tok, pos, lit = p.scanIgnoreWhitespace()
+
+	// Check for required TO token.
+	if tok != TO {
+		return nil, newParseError(tokstr(tok, lit), []string{"TO"}, pos)
+	}
```
Comment (dgnorton) [inline influxql/parser.go]: Parsing `RENAME TO` tokens can be condensed to...
```
// Parse required RENAME TO tokens.
if err := p.parseTokens([]Token{RENAME, TO}); err != nil {
   return nil, err
}
```

File: influxql/parser.go (hunk 2 above)
Comment (linearb) [inline influxql/parser.go]: mistake-- will remove.

File: influxql/parser.go (hunk 5 above)
Comment (linearb) [inline influxql/parser.go]: good point.

File: influxql/parser_test.go (hunk 3 above)
Comment (linearb) [inline influxql/parser_test.go]: yes

File: meta/internal/meta.proto (hunk 4 above)
Comment (linearb) [inline meta/internal/meta.proto]: ok - will fix.

File: meta/data.go [hunk 6]
```diff
@@ -177,6 +177,43 @@ func (data *Data) DropDatabase(name string) error {
...
+	}
+	if data.Database(oldName) == nil {
+		return ErrDatabaseNotFound
+	}
+	// find database named oldName and rename it to newName
+	for i := range data.Databases {
+		if data.Databases[i].Name == oldName {
+			data.Databases[i].Name = newName
+			data.switchDatabaseUserPrivileges(oldName, newName)
+			// TODO should rename the databases used in continuous queries
```
Comment (dgnorton) [inline meta/data.go]: @otoolep should we return an error (for now) if the DB has CQs?

Comment (linearb) [inline meta/data.go]: Won't any CQs targeting a non-existent database just fail with an error anyway?

Comment (otoolep) [inline meta/data.go]: @linearb -- I think they would fail at query time, which is quite difficult to track down in the current system. I think we should error for now, and take the conservative approach with this new command. When we improve CQ handling we could remove this constraint.

Thoughts?

Comment (linearb) [inline meta/data.go]: @otoolep  --- OK-- I see why it would be good to error at this point then.

Would it be sufficient to check all the CQs and throw the error if the 
query contains " ON <olddb> " or " INTO \"<olddb>\"."  or " FROM \"<olddb>\"." ?

I guess the error should be a new one (ErrDatabaseRenameCQConflict ?)
What do you think?

Comment (otoolep) [inline meta/data.go]: That approach sounds fine to me, and the error makes sense. I would have to see the final code to be sure.

**Comment by linearb** [comment]
@otoolep Suggested changes made.

**Comment by otoolep** [comment]
Thanks, I'll review now.

Can you rebase against master to remove the "merge" commits?

**Comment by dgnorton** [comment]
+1

**Comment by otoolep** [comment]
Thanks @linearb -- great work.

**Comment by dgnorton** [comment]
We've decided to revert this change.  See https://github.com/influxdb/influxdb/issues/4503 for details.

**Comment by linearb** [comment]
Too bad-- guess there was more involved here than any of us realized.

**Comment by dgnorton** [comment]
@linearb sorry we didn't catch this in the initial review.  We really appreciate the effort you put in on it!

**Comment by linearb** [comment]
@dgnorton @otoolep Since I think this would be a good feature to get in, I'm willing to add
something to deal with problems raised in #4503 . Before going ahead, perhaps there
might be some consensus about which approach to take (or at least which to avoid)?

**Comment by otoolep** [comment]
Thanks @linearb -- we have not reached consensus on this issue. There are some difficult issues raised, with regards to renaming opening directories. The issue with some nodes crashing before the rename completes may also be difficult. 

Perhaps some sort of simple aliasing is the best possible solution, which makes crash-recovery issue the simplest, and doesn't require any renaming of files on disk. The aliasing would be kept by Raft, and would exist in RAM only on all nodes. I believe this was suggested by @corylanou.

The mapping from alias to actual name would be referenced each time during queries and during writes. This would mean the change would touch much more of the code however.

**Comment by otoolep** [comment]
Actually, after some further discussions here, if instead the system used the database IDs as the filenames on disk, then it would be reasonably easy to support renames.

**Comment by beckettsean** [comment]
As long as there is a way for users to discover the mapping between database name and database ID that's a reasonable solution. I don't think it would be okay to obscure which database owned which files on disk.

**Comment by otoolep** [comment]
No reason why `SHOW DATABASES` wouldn't show this information, it would be easy enough.

**Comment by beckettsean** [comment]
That absolutely works for me.

**Comment by linearb** [comment]
Sounds good. I'll put something together.

**Comment by linearb** [comment]
@otoolep Unless there is an internal database ID that I'm not seeing, I'll just add one like node id, or cluster id? Also, as changing the filenames is a breaking change, let me know if you have infrastructure for upgrades/downgrades to be aware of.

**Comment by otoolep** [comment]
No, it appears we don't have IDs for databases yet. You may need to add something to the `DatabaseInfo` type.

We also don't have any explicit code for upgrades and downgrades right now. We are open to suggestions, but I realise it can be a complex topic.
